src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp

```diff
@@ -82,52 +115,42 @@ struct PrepareSolve {
         Parallel::get_parallel_component<
             ResidualMonitor<Metavariables, FieldsTag, OptionsGroup>>(cache));
 
-    return {
-        std::move(box),
-        // Terminate algorithm for now. The `ResidualMonitor` will receive the
-        // reduction that is performed above and then broadcast to the following
-        // action, which is responsible for restarting the algorithm.
-        true};
-  }
-};
-
-template <typename FieldsTag, typename OptionsGroup>
-struct InitializeHasConverged {
-  template <
-      typename ParallelComponent, typename DbTagsList, typename Metavariables,
-      typename ArrayIndex, typename DataBox = db::DataBox<DbTagsList>,
-      Requires<db::tag_is_retrievable_v<
-          LinearSolver::Tags::HasConverged<OptionsGroup>, DataBox>> = nullptr>
-  static void apply(db::DataBox<DbTagsList>& box,
-                    Parallel::GlobalCache<Metavariables>& cache,
-                    const ArrayIndex& array_index,
-                    const Convergence::HasConverged& has_converged) noexcept {
-    db::mutate<LinearSolver::Tags::HasConverged<OptionsGroup>>(
-        make_not_null(&box),
-        [&has_converged](const gsl::not_null<Convergence::HasConverged*>
-                             local_has_converged) noexcept {
-          *local_has_converged = has_converged;
-        });
-
-    // Proceed with algorithm
-    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
-        .perform_algorithm(true);
+    return {std::move(box)};
   }
 };
 
 template <typename FieldsTag, typename OptionsGroup>
-struct PrepareStep {
+struct InitializeHasConverged
+    : Receive<Tags::InitialHasConverged<OptionsGroup>,
```

Comment:
Could you try adding a

```cpp
using is_ready = Receive<..>::is_ready;
```
into the action? That would make it clear what this base class is doing. Also, maybe a clearer name than `Receive` can be thought of :)

Reply:
Here's the comment from above, since we probably typed at the same time: https://github.com/sxs-collaboration/spectre/pull/2495#discussion_r487209170
> Please just let me know how you would prefer this pattern be handled. The options I can think of are (in order of decreasing code duplication):
> 
> 1. Copy-paste the `inbox_tags`, `is_ready` function and `extract_from_inbox` function to the 6 places in this PR where they are used.
> 2. Factor them into `detail` namespaces as is currently done in this PR.
> 3. Factor them into files in `Parallel/`, as is done in #2489.

Reply:
I tried the `using is_ready = ...` pattern, but it doesn't work since `is_ready` is a function, not a type.. unless there's some C++ syntax that I missed.

Reply:
Regarding names all I could think of was `WaitUntilReceived<InboxTag, TemporalIdTag>`

Reply:
```
struct B { template<class T> void f(); };
struct D : B {
      using B::f;      // OK: names a template
//    using B::f<int>; // Error: names a template specialization
      void g() { f<int>(); }
};
```
It should be doable. I've done this before :) It's exactyl how `make_overloader` works.

Why not `SimpleIsReady` or some variant thereof? I'm looking something that indicates it's an `is_ready` supplier

Reply:
Yep that works, thanks! (no `=`)

`SimpleIsReady` doesn't read as nice, but I like that it makes clear what it is, and I can't think of anything better.

Reply:
How about `IsReadyWhenReceived<InboxTag, TemporalIdTag>`?

Reply:
That doesn't really narrow down what is checked, though. Literally any `is_ready` struct should be called that then, or am I missing something? I was hoping to have some indication that this is the simplest `is_ready` you can write "It exists". Unsure :shrug:

Reply:
Agreed. For this PR I inlined the `is_ready` method and the `extract_from_inbox` function for now, so we can move this along. I'm happy to factor them out again either in this PR or in #2489 if you or @moxcodes prefer that, but I'd like to avoid holding up this PR with that discussion.